### PR TITLE
Expand interface to support trace links

### PR DESCRIPTION
**Pull request by Dale-Koenig**
Opentelemetry supports indirect links to other spans.  Expand the interface to support this, with backwards compatibility.

### CODE REVIEW

File: alica_engine/include/engine/IAlicaTrace.h [hunk 1]
```diff
@@ -57,13 +57,34 @@ class IAlicaTrace
     }
 };
 
+class SpanLink
+{
+public:
+    std::string context;
+    // Warning:  Ensure string_views in attributes are valid for the lifetime of the span link
+    std::unordered_map<std : string, IAlicaTrace::TraceValue> attributes;
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/IAlicaTrace.h]:
```suggestion
    std::unordered_map<std::string, IAlicaTrace::TraceValue> attributes;
```

File: alica_engine/include/engine/IAlicaTrace.h [hunk 2]
```diff
@@ -57,13 +57,34 @@ class IAlicaTrace
     }
 };
 
+class SpanLink
+{
+public:
+    std::string context;
+    // Warning:  Ensure string_views in attributes are valid for the lifetime of the span link
+    std::unordered_map<std::string, IAlicaTrace::TraceValue> attributes;
+}
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/IAlicaTrace.h]:
```suggestion
};
```

File: alica_engine/include/engine/IAlicaTrace.h [hunk 3]
```diff
@@ -57,10 +58,24 @@ class IAlicaTrace
...
+    std::unordered_map<std::string, IAlicaTrace::TraceValue> attributes;
+};
+
 class IAlicaTraceFactory
 {
 public:
-    virtual std::unique_ptr<IAlicaTrace> create(const std::string& opName, std::optional<const std::string> parent = std::nullopt) const = 0;
+    // Two argument version of create.  Deprecated.
+    virtual std::unique_ptr<IAlicaTrace> create(const std::string& opName, std::optional<const std::string> parent = std::nullopt) const { return nullptr; }
+    virtual std::unique_ptr<IAlicaTrace> create(const std::string& opName, std::optional<const std::string> parent_context, const std::vector<SpanLink>&) const
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/IAlicaTrace.h]: My idea is that if we make a breaking changes release, we can remove the two arg version and add default args to the three arg version.

**Review by veerajsk — APPROVED** [review]

**Review by bjoernschroeder — APPROVED** [review]

**Comment by Dale-Koenig** [comment]
I realized this doesn't really work.  Since alica generates the traces for you, and the only interaction the user has is to create the context, to support links we must also support returning context with links included

File: alica_engine/include/engine/IAlicaTrace.h [hunk 4]
```diff
@@ -34,8 +35,7 @@ class IAlicaTrace
         {
         }
 
-    private:
```
Comment (veerajsk) [inline alica_engine/include/engine/IAlicaTrace.h:37]: why is this public now?

Comment (Dale-Koenig) [reply alica_engine/include/engine/IAlicaTrace.h:37]: Every tracing library uses their own internal variant definition, and to convert we usually need to call std::visit on the raw variant, so pretty much any tracing implementation will need access to the raw variant here.  For example to convert to AttributeValue from https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/common/attribute_value.h

Comment (Dale-Koenig) [reply alica_engine/include/engine/IAlicaTrace.h:37]: I suppose we can remove the "friend IAlicaTrace" now too

**Review by bjoernschroeder — APPROVED** [review]

**Pull request merged**